wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/WorkspaceServiceTermination.java

```diff
@@ -0,0 +1,76 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2017 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.system.server;
+
+import org.eclipse.che.api.core.model.workspace.WorkspaceStatus;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.system.shared.event.service.SystemServiceItemStoppedEvent;
+import org.eclipse.che.api.system.shared.event.service.SystemServiceStoppedEvent;
+import org.eclipse.che.api.workspace.server.WorkspaceManager;
+import org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent;
+
+import javax.inject.Inject;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Terminates workspace service.
+ *
+ * @author Yevhenii Voevodin
+ */
+class WorkspaceServiceTermination implements ServiceTermination {
+
+    @Inject
+    private WorkspaceManager workspaceManager;
+
+    @Inject
+    private EventService eventService;
+
+    @Override
+    public void terminate() throws InterruptedException {
+        EventSubscriber propagator = new WorkspaceStoppedEventsPropagator();
+        eventService.subscribe(propagator);
+        try {
+            workspaceManager.shutdown();
+        } finally {
+            eventService.unsubscribe(propagator);
+        }
+    }
+
+    @Override
+    public String getServiceName() {
+        return "workspace";
+    }
+
+    /**
+     * Propagates workspace stopped events as {@link SystemServiceStoppedEvent} events.
+     */
+    private class WorkspaceStoppedEventsPropagator implements EventSubscriber<WorkspaceStatusEvent> {
+
+        private final int           totalRunning;
+        private final AtomicInteger currentlyStopped;
+
+        private WorkspaceStoppedEventsPropagator() {
+            this.totalRunning = workspaceManager.getRunningWorkspacesIds().size();
```

Comment:
I consider running as not stopped, because if it's running you can't start another one, which is true for starting/snapshotting/stopping workspaces. WorkspaceRuntimes provides methods like
`isAnyRunning`, `hasRuntime`, `getRuntimesIds` that work exactly in the same context. Do you have any proposal?